domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt

```diff
@@ -0,0 +1,105 @@
+package org.oppia.domain.oppialogger.loguploader
+
+import android.content.Context
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import org.oppia.domain.oppialogger.analytics.AnalyticsController
+import org.oppia.domain.oppialogger.exceptions.ExceptionsController
+import org.oppia.domain.oppialogger.exceptions.toException
+import org.oppia.util.logging.ConsoleLogger
+import org.oppia.util.logging.EventLogger
+import org.oppia.util.logging.ExceptionLogger
+import org.oppia.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+
+/** Worker class that extracts log reports from the cache store and logs them to the remote service. */
+class LogUploadWorker private constructor(
+  context: Context,
+  params: WorkerParameters,
+  private val analyticsController: AnalyticsController,
+  private val exceptionsController: ExceptionsController,
+  private val exceptionLogger: ExceptionLogger,
+  private val eventLogger: EventLogger,
+  private val consoleLogger: ConsoleLogger,
+  @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+) : CoroutineWorker(context, params) {
+
+  companion object {
+    const val WORKER_CASE_KEY = "worker_case_key"
+    const val TAG = "LogUploadWorker.tag"
+    const val EVENT_WORKER = "event_worker"
+    const val EXCEPTION_WORKER = "exception_worker"
+  }
+
+  override suspend fun doWork(): Result {
+    return when (inputData.getString(WORKER_CASE_KEY)) {
+      EVENT_WORKER -> {
+        withContext(backgroundDispatcher) { uploadEvents() }
+      }
+      EXCEPTION_WORKER -> {
+        withContext(backgroundDispatcher) { uploadExceptions() }
+      }
+      else -> Result.failure()
+    }
+  }
+
+  /** Extracts exception logs from the cache store and logs them to the remote service. */
+  private suspend fun uploadExceptions(): Result {
+    return try {
+      val exceptionLogs = exceptionsController.getExceptionLogStoreList()
+      exceptionLogs.let {
+        for (exceptionLog in it) {
+          exceptionLogger.logException(exceptionLog.toException())
+          exceptionsController.removeFirstExceptionLogFromStore()
+        }
+      }
+      Result.success()
```

Comment:
What if the log operation fails? Should we track it & return a failure in that case, then retry later?

Reply:
If any of the operation fails, it'll throw an error which will be caught and the result will be failed. All of the un-uploaded logs will then be taken up in the next work manager cycle. Does that make sense?

Reply:
I don't think that's quite right, unless I'm missing something. The suspend function you're calling on the DataProvider (retrieveData()) won't throw an exception, it'll return an AsyncResult with the failure. Don't we need to check all of these results for potential failures?

Reply:
Umm..I'm actually using the readDataAsync from the Analytics/Exceptions Controller. Will that not throw an error ? I believe I faced an error while testing.

Reply:
No. See the implementation: https://github.com/oppia/oppia-android/blob/2ca2397/data/src/main/java/org/oppia/data/persistence/PersistentCacheStore.kt#L75.

Reply:
Done.

Reply:
Resolving this now.